lib/btree.js

```diff
@@ -0,0 +1,245 @@
+    // the biggest element in the node has value === undefined, therefore,
+    // in order to get true for it here using reverse comparison.
+    else if (!(value >= itemValue)) end = i - 1;
+    else return [true, i];
+  }
+  return [false, start];
+};
+
+function* inorderTraversal(start, finish, currNode) {
+  const binarySearchResult = binarySearch(currNode, start);
+  let startIndex = binarySearchResult[1];
+  if (binarySearchResult[0]) startIndex++;
+  const finishIndex = finish === undefined ?
+    currNode.length - 1 :
+    binarySearch(currNode, finish)[1];
+  if (isLeaf(currNode)) {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      yield currNode[currIndex].data;
+    }
+  } else {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      const currElement = currNode[currIndex];
+      yield* inorderTraversal(start, finish, currElement.child);
+      yield currElement.data;
+    }
+    yield* inorderTraversal(start, finish, currNode[finishIndex].child);
+  }
+}
+
+const joinNodes = (parent, firstNodeIndex, secondNodeIndex) => {
+  const firstNode = parent[firstNodeIndex].child;
+  const secondNode = parent[secondNodeIndex].child;
+  const mid = parent.splice(firstNodeIndex, 1)[0];
+  mid.child = firstNode.pop().child;
+  secondNode.splice(0, 0, ...firstNode, mid);
+  return secondNode;
+};
+
+const growChild = (parent, childIndex) => {
+  const parentElement = parent[childIndex];
+  const node = parentElement.child;
+  const minDegree = node.length;
+  if (childIndex > 0) { // If we have left neighbor
+    const leftNeighbor = parent[childIndex - 1].child;
+    if (leftNeighbor.length > minDegree) {
+      const extractedElement =
+        leftNeighbor.splice(leftNeighbor.length - 2, 1)[0];
+      const insertedElement = new Element(
+        parent[childIndex - 1].value,
+        parent[childIndex - 1].data,
+        leftNeighbor[leftNeighbor.length - 1].child
+      );
+      leftNeighbor[leftNeighbor.length - 1].child = extractedElement.child;
+      extractedElement.child = leftNeighbor;
+      parent.splice(childIndex - 1, 1, extractedElement);
+      node.unshift(insertedElement);
+      return node;
+    }
+  }
+  if (childIndex < parent.length - 1) { // If we have right neighbor
+    const rightNeighbor = parent[childIndex + 1].child;
+    if (rightNeighbor.length > minDegree) {
+      const extractedElement = rightNeighbor.shift();
+      const insertedElement = new Element(
+        parentElement.value,
+        parentElement.data,
+        node[node.length - 1].child
+      );
+      node[node.length - 1].child = extractedElement.child;
+      extractedElement.child = node;
+      parent.splice(childIndex, 1, extractedElement);
+      node.splice(node.length - 1, 0, insertedElement);
+      return node;
+    }
+    return joinNodes(parent, childIndex, childIndex + 1);
+  }
+  return joinNodes(parent, childIndex - 1, childIndex);
```

Comment:
Not really, this code works when we don't have a left neighbor, and when the length of the right neighbor is equal to bTree.minDegree
![1](https://user-images.githubusercontent.com/31672085/45171549-5d082780-b20b-11e8-8e94-2342672c920c.PNG)

Reply:
Also, if i correctly understand, how ``nyc`` works, you can see that tests enter this block, and work out correctly:
![image_2018-09-06_19-28-23](https://user-images.githubusercontent.com/31672085/45171717-cb4cea00-b20b-11e8-8023-7da1c6898f4b.png)

Reply:
:thinking: According to your decision tree:
If we go right->right then we don't have left or right neighbor but we 'join out node with left node'?

Reply:
Not all routes in this decision tree are achievable. We can't go to the right ->right, since at least one neighbor we must have, this is embedded in the rules of the b-tree. If we go to the right in the beginning, then we must go to the left. I did not put this in a separate block, because it seemed to me that too much code would be repeated, or it would be necessary to create additional functions. What do you think, which option is acceptable?

Reply:
We can get to the block ``if`` on L97 from two places: from if on L80 and L82. Accordingly, from L80 we can't get to L114, but for L82 we can. That is, if we have a left neighbor, but it is too small to extract an element from it, we will get to L97, and then, if we have a right neighbor, we'll get on L114.

Reply:
You mean ' if we have a right neighbor, we'll get on L114.' if we **don't** have a right neighbor?